WordPress/openverse-frontend
# Add local visual regression infrastructure

## Fixes


Fixes #1017 by @sarayourfriend
Fixes #1019 by @sarayourfriend

## Description


Changes in this PR:
* Renames the `e2e` test suite to `playwright`
* Adds a folder of visual regression tests to the playwright test suite
* Splits the existing playwright test utils into two separate modules and adds a new `breakpoints` utils for simplifying running a single scenario against multiple breakpoints
* Removes the half-baked cross-platform support and in the process removes playwright runner script duplication ~allowing us to rely entirely on the package.json scripts~
    * We still have to use a `bin/playwright.sh` in order to easily support passing arguments directly to `playwright` in the docker container.
* Fixes the docker tapes permission issue (thanks to dropping native Windows support it is possible to use `id -u` to get the user ID and have the correct permissions from the get go)
* Updates CI to run the full playwright suite
* Adds alias module resolution to Playwright. This prevents us from having to use relative imports both from `src` and from the root directory. **Should we enable the import rule to not allow relative imports?** :thinking: 
* Moves the Playwright testing information out of TESTING_GUIDELINES.md into a `test/playwright/README.md` and expands the documentation to include information about visual regression testing as well as additional information about the Talkback proxy and how it is configured.
* Uploads failed test-result artifacts. This not only includes the `actual` `expected` and `diff` snapshots but also the `trace.zip` for _any_ failed Playwright test, meaning it should be easier to debug all flaky Playwright tests.

Notably:
* Does not use `jest-image-snapshot`. This library doesn't support playwright/test, we'd have to switch to playwright-jest which we previously decided against using (which I think is fine). It also turned out to be unnecessary as Playwright does actually generate the `expected` `actual` and `diff` versions of the snapshot for failed tests in `test-results` whenever you pass a file extension to the snapshot name.
* Paves the way to setup storybook playwright tests (#1020). For that PR we'll move `playwright.config.ts` up one level and rename it `playwright.config.base.ts` and be able to just change a few properties to fit running the storybook in the playwright container instead of the Nuxt app

For the storybook PR, there's two approaches I think that follow naturally from this PR:
1. Either setup storybook in a separate folder (`test/storybook/playwright.config.ts`);
2. or else sub-folder the `playwright` suite further (`test/playwright/{storybook,nuxt}/playwright.config.ts`)

I think I like the second one better. Either way we'd want to update the package.json scripts to be clearer... `test:playwright` -> `test:playwright:nuxt`? It gets verbose if we do it that way. We could shorten playwright to `pw`, `test:pw:nuxt`, `test:pw:storybook` isn't so bad.

## Testing Instructions

Checkout this branch and run `pnpm install`.

Run `pnpm test:playwright` and check that the tests pass. Run the other `test:playwright:*` scripts and verify they do all the things we'd expect them to do and work as expected.

Also update something on the homepage (add a new bit of text or change the background color) and run `pnpm test:playwright -- visual-regression` and observe how failures are handled, including the `diff` generation in `test-results`. Likewise run `pnpm test:playwright -- visual-regression -u` to test snapshot updating.

~NOTE: Currently the snapshot tests aren't passing and I still need to debug what's going on there.~

It turns out I forgot to add the `data-testid` attribute that was necessary for deleting the image carousel from the page during snapshot tests. The image carousel is random from page-load to page-load so it introduces variations that cause tests to be flaky. It'd be nice if we could configure it to be non-random (perhaps using a feature flag? :eyes: https://github.com/WordPress/openverse/issues/178) but for now just deleting it from the page simplifies it.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
